### PR TITLE
Bug 1576223 - Testing a better image and card load

### PR DESCRIPTION
**Pull request by ScottDowne**
To test:

You should see roughly the same experience before and after the patch as these videos:
https://drive.google.com/file/d/1V8R167fe2grm3vIiYdERCqIEE8QtCdBo/view?usp=sharing
https://drive.google.com/file/d/1fw2kCaJUxrGyhZQZ8FH2cAuRb1qnOFVZ/view?usp=sharing

### CODE REVIEW

**Comment by ScottDowne** [comment]
Oh I forgot to look at tests for this, probably failing.

File: content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx [hunk 1]
```diff
@@ -50,7 +59,15 @@ export class DSImage extends React.PureComponent {
...
   componentDidMount() {
-    this.observer = new IntersectionObserver(this.onSeen.bind(this));
+    this.idleCallbackId = window.requestIdleCallback(
+      this.onIdleCallback.bind(this)
+    );
+    this.observer = new IntersectionObserver(this.onSeen.bind(this), {
+      // Assume an image will be eventually seen if it is within 520px of the viewport
+      // This is half the average Desktop vertical screen size:
+      // http://gs.statcounter.com/screen-resolution-stats/desktop/north-america
+      rootMargin: `540px`,
```
Comment (punamdahiya) [inline content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx]: is it 520 or 540px, should update comment above to say 540px

Comment (gvn) [reply content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx]: 540 👍

File: content-src/components/DiscoveryStreamComponents/DSCard/DSCard.jsx [hunk 2]
```diff
@@ -129,7 +129,21 @@ export class DSCard extends React.PureComponent {
...
+    if (!this.state.isSeen) {
+      if (this.observer && this.placholderElement) {
+        this.observer.unobserve(this.placholderElement);
+      }
+      this.setState({
+        isSeen: true,
+      });
+    }
+  }
+
```
Comment (punamdahiya) [inline content-src/components/DiscoveryStreamComponents/DSCard/DSCard.jsx]: Change above is loading all 21 cards and its content on idle after first load.

File: content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx [hunk 3]
```diff
@@ -50,7 +59,15 @@ export class DSImage extends React.PureComponent {
...
   componentDidMount() {
-    this.observer = new IntersectionObserver(this.onSeen.bind(this));
+    this.idleCallbackId = window.requestIdleCallback(
+      this.onIdleCallback.bind(this)
+    );
+    this.observer = new IntersectionObserver(this.onSeen.bind(this), {
+      // Assume an image will be eventually seen if it is within 520px of the viewport
+      // This is half the average Desktop vertical screen size:
+      // http://gs.statcounter.com/screen-resolution-stats/desktop/north-america
+      rootMargin: `520px`,
```
Comment (punamdahiya) [inline content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx]: With this change noticed first nine cards images loaded on first load and the rest with use-transition class.

**Comment by punamdahiya** [comment]
Verified and no visible issues found with the patch, it does improve experience by taking away a very slight transition seen from placeholder to content while scrolling. However above two observation/comments in patch might have some performance expense and will be good to overrule that.

**Comment by ScottDowne** [comment]
Numbers for this: https://docs.google.com/document/d/1PoiF9A70RdEQKe3B1q_2blmy0GAlpeR41Piq3wM_t7Q/edit#

File: content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx (hunk 3 above)
Comment (gvn) [reply content-src/components/DiscoveryStreamComponents/DSImage/DSImage.jsx]: Should be 540

**Pull request merged**